builtin/providers/cloudstack/resource_cloudstack_instance.go

```diff
@@ -255,6 +265,32 @@ func resourceCloudStackInstanceUpdate(d *schema.ResourceData, meta interface{})
 		d.SetPartial("service_offering")
 	}
 
+	if d.HasChange("keypair") {
+		log.Printf("[DEBUG] SSH keypair changed for %s, starting update", name)
+
+		p := cs.SSH.NewResetSSHKeyForVirtualMachineParams(d.Id(), d.Get("keypair").(string))
+		// Before we can actually change the keypair, the virtual machine must be stopped
+		_, err := cs.VirtualMachine.StopVirtualMachine(cs.VirtualMachine.NewStopVirtualMachineParams(d.Id()))
```

Comment:
Wouldn't it be better to join the changing of the keypair together with any possible changes of the service offering? As both these options need the VM to be stopped and started and if you change both it would currently do this twice for one `apply`...

So I guess something like this would be more effecient:

```go
	// Check if the service offering is changed and if so, update the offering
	if d.HasChange("service_offering") || d.HasChange("keypair"){
		// Before we can actually make these changes, the virtual machine must be stopped
		_, err := cs.VirtualMachine.StopVirtualMachine(cs.VirtualMachine.NewStopVirtualMachineParams(d.Id()))
		if err != nil {
			return fmt.Errorf(
				"Error stopping instance %s before making changes: %s", name, err)
		}

                if d.HasChange("service_offering") {
				log.Printf("[DEBUG] Service offering changed for %s, starting update", name)

				// Retrieve the service_offering UUID
				serviceofferingid, e := retrieveUUID(cs, "service_offering", d.Get("service_offering").(string))
				if e != nil {
					return e.Error()
				}

				// Create a new parameter struct
				p := cs.VirtualMachine.NewChangeServiceForVirtualMachineParams(d.Id(), serviceofferingid)


				// Change the service offering
				_, err = cs.VirtualMachine.ChangeServiceForVirtualMachine(p)
				if err != nil {
					return fmt.Errorf(
						"Error changing the service offering for instance %s: %s", name, err)
				}

				d.SetPartial("service_offering")
                }

                if d.HasChange("keypair") {
		                log.Printf("[DEBUG] SSH keypair changed for %s, starting update", name)

		                // Create a new parameter struct
		                p := cs.SSH.NewResetSSHKeyForVirtualMachineParams(d.Id(), d.Get("keypair").(string))

				// Change the ssh keypair
				_, err = cs.SSH.ResetSSHKeyForVirtualMachine(p)
				if err != nil {
					return fmt.Errorf(
						"Error changing the SSH keypair for instance %s: %s", name, err)
				}

				d.SetPartial("keypair")
		}

		// Start the virtual machine again
		_, err = cs.VirtualMachine.StartVirtualMachine(cs.VirtualMachine.NewStartVirtualMachineParams(d.Id()))
		if err != nil {
			return fmt.Errorf(
				"Error starting instance %s after making changes: %s", name, err)
		}
	}
```